redbeams/src/main/java/com/sequenceiq/redbeams/service/network/SubnetChooserService.java

```diff
@@ -7,38 +7,41 @@
 
 import org.springframework.stereotype.Service;
 
+import com.sequenceiq.cloudbreak.cloud.init.CloudPlatformConnectors;
+import com.sequenceiq.cloudbreak.cloud.model.CloudPlatformVariant;
 import com.sequenceiq.cloudbreak.cloud.model.CloudSubnet;
+import com.sequenceiq.cloudbreak.cloud.model.SubnetSelectionParameters;
 import com.sequenceiq.cloudbreak.common.mappable.CloudPlatform;
-import com.sequenceiq.redbeams.exception.BadRequestException;
-import com.sequenceiq.redbeams.exception.RedbeamsException;
+import com.sequenceiq.redbeams.api.endpoint.v4.stacks.aws.AwsDatabaseServerV4Parameters;
+import com.sequenceiq.redbeams.domain.stack.DBStack;
 
 @Service
 public class SubnetChooserService {
 
     @Inject
-    private AwsSubnetChooser awsSubnetChooser;
-
-    public List<CloudSubnet> chooseSubnets(List<CloudSubnet> subnetMetas, CloudPlatform cloudPlatform, Map<String, String> dbParameters) {
-        switch (cloudPlatform) {
-            case AWS:
-                return awsSubnetChooser.chooseSubnets(subnetMetas, dbParameters);
-            case AZURE:
-                return chooseSubnetsAzure(subnetMetas);
-            case MOCK:
-                return chooseSubnetsMock(subnetMetas);
-            default:
-                throw new RedbeamsException(String.format("Support for cloud platform %s not yet added", cloudPlatform.name()));
-        }
+    private CloudPlatformConnectors cloudPlatformConnectors;
+
+    public List<CloudSubnet> chooseSubnets(List<CloudSubnet> subnetMetas, CloudPlatform cloudPlatform, DBStack dbStack) {
+        return cloudPlatformConnectors.get(new CloudPlatformVariant(dbStack.getCloudPlatform(), dbStack.getPlatformVariant()))
+                .networkConnector()
+                .selectSubnets(
+                        subnetMetas,
+                        SubnetSelectionParameters.builder()
+                                .withHa(isHa(cloudPlatform, dbStack))
+                                .withPreferPrivateNetwork()
+                                .build()
+                );
     }
 
-    private List<CloudSubnet> chooseSubnetsAzure(List<CloudSubnet> subnetMetas) {
-        if (subnetMetas.isEmpty()) {
-            throw new BadRequestException("Insufficient number of subnets: at least one subnet needed");
+    private boolean isHa(CloudPlatform cloudPlatform, DBStack dbStack) {
+        boolean ha = true;
+        if (CloudPlatform.AWS.equals(cloudPlatform)) {
+            ha = !isAwsNoHaRequested(dbStack.getParameters());
         }
-        return subnetMetas.subList(0, 1);
+        return ha;
     }
```

Comment:
I would move this logic into `DBStack` as it could be reused there.
also the `CloudPlatform` parameter is not necessary, as it's in `dbStack`

Reply:
Great idea, this belongs there. Did it.
On the other hand, cloudPlatform in DBStack is a string. I did not want to convert again.